src/misc/8.md

```diff
@@ -0,0 +1,42 @@
+# Misc 8 @adotinthevoid
+
+{{#include ../include/quiz-is-wip.md}}
+
+```rust
+{{#include ../../code/examples/misc_8.rs:matters}}
+```
+
+<details>
+<summary>Solution</summary>
+
+```
+{{#include ../../code/examples/stderr/misc_8.stderr}}
+```
+
+There are 3 kinds of structs in rust:
+
+1. Plain structs (eg `struct Foo{bar: i32}`)
+2. Tuple structs (eg `struct Bar(i32);`)
+3. Unit structs (eg `struct Baz;`)
+
+However all 3 kinds can have a struct with no fields.
+
+Types and values live in seperate namespaces, as it is usually [^not1] possible to
+syntacticly determine weather an identifier the compiller needs to resolve will
+be an value or type .
+
+[^not1]: except when it isn't :)
+
+This means that `struct bar` only gets inserted into the type namespace, and `fn
+bar` only gets inserted to the value namespace (as functions in rust are first
+class values [^not]). Therefor both of these declarations can co-exist.
```

Comment:
```suggestion
class values [^not]). Therefore both of these declarations can co-exist.
```
when are they not?

Reply:
Because they can't be higher ranked wrt type params.

IE I can write

```rust
    fn identity<T>(t: T) -> T {
        t
    }
    
    // Ok
    let x: fn(i32) -> i32 = identity;
```

but not

```rust
    let y: for<T> fn(T) -> T = identity;
```

Not sure if this is worth including.

Reply:
Whereas in haskell you can do

```haskell
foo :: forall a b. (forall t. [t] -> t) -> [a] -> [b] -> (a, b)
foo sel as bs = (sel as, sel bs)
```

or to be higher kinded also:

```haskell
foo :: forall a b f. (forall t. t -> f t) -> a -> b -> (f a, f b)
foo sel a b = (sel a, sel b)
```